modules/appgw/variables.tf

```diff
@@ -1,169 +1,632 @@
+      name = "minimum"
+    }
+  }
+  nullable = false
+  type = map(object({
+    name                  = optional(string)
+    path                  = optional(string)
+    hostname_from_backend = optional(bool, false)
+    hostname              = optional(string)
+    port                  = optional(number, 80)
+    protocol              = optional(string, "Http")
+    timeout               = optional(number, 60)
+    cookie_based_affinity = optional(string, "Enabled")
+    affinity_cookie_name  = optional(string)
+    probe                 = optional(string)
+    root_certs = optional(map(object({
+      name = string
+      path = string
+    })), {})
+  }))
+  validation {
+    condition = alltrue(flatten([
+      for _, backend in var.backends : [
+        contains(["Http", "Https"], backend.protocol)
+    ]]))
+    error_message = "Possible values for `protocol` are `Http` and `Https`."
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, backend in var.backends : [
+        contains(["Enabled", "Disabled"], backend.cookie_based_affinity)
+    ]]))
+    error_message = "Possible values for `cookie_based_affinity` are Enabled and Disabled."
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, backend in var.backends : (backend.port >= 1 && backend.port <= 65535)
+    ]))
+    error_message = "The backend `port` should be a valid TCP port number from 1 to 65535."
+  }
+  validation {
+    condition = alltrue(flatten([
+      for _, backend in var.backends : (backend.timeout != null ? backend.timeout >= 1 && backend.timeout <= 86400 : true)
+    ]))
+    error_message = "The backend `timeout` property should can take values between 1 and 86400 (seconds)."
+  }
+  validation {
+    condition     = length(flatten([for _, backend in var.backends : backend.name])) == length(distinct(flatten([for _, backend in var.backends : backend.name])))
+    error_message = "The `name` property has to be unique among all backends."
+  }
 }
 
-variable "ssl_policy_name" {
+variable "probes" {
   description = <<-EOF
-  Name of an SSL policy. Supported only for `ssl_policy_type` set to `Predefined`. Normally you can set it also for `Custom` policies but the name is discarded on Azure side causing an update to Application Gateway each time terraform code is run. Therefore this property is omitted in the code for `Custom` policies. 
-  
-  For the `Predefined` polcies, check the [Microsoft documentation](https://docs.microsoft.com/en-us/azure/application-gateway/application-gateway-ssl-policy-overview) for possible values as they tend to change over time. The default value is currently (Q1 2022) a Microsoft's default.
+  A map of probes for the Application Gateway.
+
+  Every probe contains attributes:
+  - `name`       - (`string`, required) The name used for this Probe
+  - `path`       - (`string`, required) The path used for this Probe
+  - `host`       - (`string`, optional) The hostname used for this Probe
```

Comment:
We had a conversation with @acelebanski about this setting. I think we should add a default of `127.0.0.1` here. Otherwise, when you define a backed and a probe with just the defaults, the code will fail. The other option is to investigate again a possibility to add NICs to the backend pool. But it has to be tested with a public Load Balancer hooked to the same NICs. If this would work, then we could skip a default for this setting and set `backends.hostname_from_backend = true` (this would work with VMSS as well).

Reply:
I extended validation to 3rd option - pick host name from backend target:

https://github.com/PaloAltoNetworks/terraform-azurerm-vmseries-modules/blob/e78e0d91cbba7ad5d16e95980aba44807b0335dd/modules/appgw/main.tf#L323C1-L335

Unfortunately in backend pool for application gateway there is no possibility to use NIC, so we are using IPs for VMSS.